res/controllers/Gemini-G4V-scripts.js

```diff
@@ -0,0 +1,562 @@
+    engine.setParameter("[EffectRack1_EffectUnit2_Effect1]", "enabled", 1);
+    engine.setParameter("[EffectRack1_EffectUnit3_Effect1]", "enabled", 1);
+    engine.setParameter("[EffectRack1_EffectUnit4_Effect1]", "enabled", 1);
+
+};
+
+g4v.shutdown = function()  {
+    // Shutdown components
+    g4v.otherControls.shutdown();
+    g4v.deck1Mixer.shutdown();
+    g4v.deck1Mixer.shutdown();
+    g4v.deck2Mixer.shutdown();
+    g4v.deck3Mixer.shutdown();
+    g4v.deck4Mixer.shutdown();
+    g4v.leftDeck.shutdown();
+    g4v.rightDeck.shutdown();
+
+    // Switch off leds not connected to controls
+    midi.sendShortMsg(0x90, 0x1B, 0x00);	// Left HotCue pad mode select button
+    midi.sendShortMsg(0x90, 0x1C, 0x00);	// Left Auto Loop pad mode select button
+    midi.sendShortMsg(0x90, 0x1D, 0x00);	// Left Sample pad mode select button
+    midi.sendShortMsg(0x90, 0x1E, 0x00);	// Left Loop Roll pad mode select button
+    midi.sendShortMsg(0x90, 0x26, 0x00);	// Left Deck select button
+    midi.sendShortMsg(0x91, 0x1B, 0x00);	// Right HotCue pad mode select button
+    midi.sendShortMsg(0x91, 0x1C, 0x00);	// Right Auto Loop pad mode select button
+    midi.sendShortMsg(0x91, 0x1D, 0x00);	// Right Sample pad mode select button
+    midi.sendShortMsg(0x91, 0x1E, 0x00);	// Right Loop Roll pad mode select button
+    midi.sendShortMsg(0x91, 0x26, 0x00);	// Right Deck select button
+
+    // Switch off VU metters
+    midi.sendShortMsg(0xB3, 0x14, 0x00);
+    midi.sendShortMsg(0xB3, 0x15, 0x00);
+    midi.sendShortMsg(0xB3, 0x16, 0x00);
+    midi.sendShortMsg(0xB3, 0x17, 0x00);
+    midi.sendShortMsg(0xB3, 0x18, 0x00);
+    midi.sendShortMsg(0xB3, 0x19, 0x00);
+};
+
+// Custom deck definition
+g4v.Deck = function(deckNumbers, midiChannel) {
+    components.Deck.call(this, deckNumbers);
+    this.cueBtn = new components.CueButton([0x8F + midiChannel, 0x02]);
+    this.cupBtn = new 	components.Button({
+        midi: [0x8F + midiChannel, 0x03],
+        inKey: "cue_gotoandplay",
+    });
+    this.playBtn = new components.PlayButton([0x8F + midiChannel, 0x01]);
+    this.syncBtn = new components.SyncButton([0x8F + midiChannel, 0x04]);
+    this.tempoPot = new components.Pot({
+        midi: [0x8F + midiChannel, 0x01],
+        inKey: "rate",
+        shift: function() { this.inKey = "pitch"; },
+        unshift: function() { this.inKey = "rate"; },
+    });
+    this.keyLockBtn = new components.Button({
+        midi: [0x8F + midiChannel, 0x05],
+        key: "keylock",
+        type: components.Button.prototype.types.toggle,
+    });
+    this.quantizeBtn = new components.Button({
+        midi: [0x8F + midiChannel, 0x07],
+        key: "quantize",
+        type: components.Button.prototype.types.toggle,
+    });
+    this.slipBtn = new components.Button({
+        midi: [0x8F + midiChannel, 0x19],
+        key: "slip_enabled",
+        type: components.Button.prototype.types.toggle,
+        deck: this,
```

Comment:
not an ideal solution. The general pattern is to alias `this` somewhere in the constructor. So add something like `const theDeck = this` at the top of the function and then you can access the current deck using `theDeck.currentDeck` below.

Reply:
I already tried something similar. But theDeck is going to be a variable in the context of the deck constructor.
When  the slip button is pressed a functions is called and is not having to have theDeck in his context and will fail. That's why I decided to store this (the deck for that button) into the slipButton object itself.
And it was working. After applying the suggested change is not working anymore.
What can I do then?

Reply:
This doesn't seem right. Functions in JS are closures, they capture references to all variables referred within them, even when coming from the outer scope. Only the scope in they were declared matters, not in which they will be called.

You must've made some other mistake since there are other mappings where this works flawlessly: https://github.com/mixxxdj/mixxx/blob/main/res/controllers/Numark-N4-scripts.js#L377